tests: Migrate from `flake8` and `pylint` to ruff

0xquark opened the pull request:
fixes #7467 
<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

0xquark committed:
tests: Replace flake8 and pylint with ruff for linting; fixes #7467

- Removed .flake8 and pylintrc configuration files.
- Moved the flake8 exclusion rules (E731, E722) into ruff, excluding W503, and set up pylint rules in ruff using PLE with default settings.
- Updated run_tests.sh to use ruff for linting instead of flake8 and pylint.

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

rdimaio reviewed (commented):
Script looks good, but we should also ensure that ruff is replacing all the checks that we're currently performing with flake8 and pylint. We should then also remove the cfg files for these latter two tools, as they are then not necessary anymore.

These two tools are configured here:
- https://github.com/rucio/rucio/blob/master/.flake8
- https://github.com/rucio/rucio/blob/master/pylintrc

And ruff is configured here: https://github.com/rucio/rucio/blob/master/pyproject.toml#L19

Could you please:
- ensure the ruff checks roughly correspond to the checks enabled in the two existing tools, as well as ignoring the errors/warnings that are set to be ignored in those tools (might be helpful: https://docs.astral.sh/ruff/faq/#how-does-ruffs-linter-compare-to-flake8)
- remove the flake8 cfg: https://github.com/rucio/rucio/blob/master/.flake8
- remove the pylint cfg: https://github.com/rucio/rucio/blob/master/pylintrc

0xquark commented:
Hi @rdimaio ,
When I run `pylint bin/* lib/ tools/*.py` on my branch locally, I get over 600 linting errors/warnings based on our [pylintrc](https://github.com/rucio/rucio/blob/master/pylintrc). It looks like we're not currently running Pylint in our CI pipeline which would also fail in this case.
I'm working on migrating from Pylint to Ruff, and if I apply the same rules, I'll get the same errors. Do you know if we're just using the default Pylintrc and not actively maintaining our code based on that?
<img width="1503" alt="image" src="https://github.com/user-attachments/assets/31f5d190-7b4a-4d42-abe6-e2093ec866ec" />

rdimaio commented:
> Hi @rdimaio , When I run `pylint bin/* lib/ tools/*.py` on my branch locally, I get over 600 linting errors/warnings based on our [pylintrc](https://github.com/rucio/rucio/blob/master/pylintrc). It looks like we're not currently running Pylint in our CI pipeline which would also fail in this case. I'm working on migrating from Pylint to Ruff, and if I apply the same rules, I'll get the same errors. Do you know if we're just using the default Pylintrc and not actively maintaining our code based on that? <img alt="image" width="1503" src="https://private-user-images.githubusercontent.com/84588720/421862276-31f5d190-7b4a-4d42-abe6-e2093ec866ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3ODczNzgsIm5iZiI6MTc0MTc4NzA3OCwicGF0aCI6Ii84NDU4ODcyMC80MjE4NjIyNzYtMzFmNWQxOTAtN2I0YS00ZDQyLWFiZTYtZTIwOTNlYzg2NmVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEyVDEzNDQzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmYWY5ZDU4YTNjNmNkMzI5Y2E5ZWMyYmMyZjQ3NmVhMjI4MjhmNzU0Nzk1YWI4OTRmNTQwMTIwNWQ0NGIxNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Hga0Pg9vjQJdcFI1gL1PHkLO2qRDgwLTTo-2vmYQWjI">

We set 'syntax' as one of the values in the test matrix here https://github.com/rucio/rucio/blob/62fd7f40ee0ac50becf0866401b00d642b39e6c0/etc/docker/test/matrix.yml

And the `pylintrc` is passed here: https://github.com/rucio/rucio/blob/62fd7f40ee0ac50becf0866401b00d642b39e6c0/tools/test/check_syntax.sh#L45

Given that these errors don't come up in the CI at the moment, I'd be in favour of just removing it all anyway, as clearly there's something wrong. I think we would benefit by going ahead with this change, moving just to ruff, and implementing these checks (which don't seem to be in effect at the moment) as we go

0xquark committed:
tests: Refactor Ruff exclusions and cleanup testing setup

- Updated ignoretool.py: Replaced pylint and flake8 exclusion of dir, files with ruff. Fixed the output format for Ruff exclusions to enclose directory paths in quotes, ensuring correct argument parsing in Ruff checks.
- Modified test.sh: Replaced Flake8 and Pylint in the TestSuite with ruff, ensuring the generated Ruff exclusion list is correctly
passed to SYNTAX_RUFF_ARGS, along with the appropriate test files.
- Refined check_syntax.sh: Adjusted logic to handle SYNTAX_RUFF_ARGS
properly during Ruff syntax checks, replacing flake8 and pylint.
- Dockerfile update: Removed the .flake8, .pylintrc file from the Docker image copy command and added pyproject.toml to fetch the ruff config in the CI job.
- Updated the requirements.dev.in/txt to include ruff and resolve any merge conflicts

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

0xquark committed:
tests: Remove warnings/linting errors in ruff

- Removed warnings generated by ruff by replacing pylint codes with ruff codes (#NOQA: F841, ARG001)
- Fixed linting errors related to E714(not-is-test), E721(type-comparison), E713(not-in-test), I001(unsorted-imports)
- chore: Add an extra blank line to `check_syntax.sh` and `requirements.dev.in`
- chore: Refix Indentation and alignment in requirements.dev.in and requirements.dev.txt
- Resolve Merge Conflict with master

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

0xquark commented:
I’ve switched from using `flake8` and `pylint` to Ruff. For Flake8, everything works in Ruff except for W503, which we are ignoring in our Flake8 config, so it’s not an issue. You can see this in the config here.
https://github.com/rucio/rucio/blob/c7b2fe9d4475c3c1ef8f52f9359f7f892c9aad58/.flake8#L3

For Pylint, our old `pylintrc` wasn’t running properly in CI, so we couldn’t detect linting errors (I found over 600 errors when I ran it locally). I’ve enabled the default Pylint settings in the Ruff config 
https://github.com/rucio/rucio/blob/8906c23585348bcc58e764d1d88f50e4a923b69e/pyproject.toml#L64
and fixed some linting errors according to these rules, as shown in this [commit](https://github.com/rucio/rucio/pull/7481/commits/8906c23585348bcc58e764d1d88f50e4a923b69e) (which weren’t many).

I've opened a separate [issue](https://github.com/rucio/rucio/issues/7499) to discuss enabling more rules in our Ruff config for Pylint and handling each rule in a separate PR.

rdimaio commented on requirements/requirements.dev.txt:
```diff
@@ -455,12 +455,14 @@ rsa==4.9
-s3transfer==0.11.4
+        ruff==0.9.10
+            # via -r requirements.dev.in
+        s3transfer==0.11.4
```
please fix the indenting here

0xquark replied:
Thanks! I would fix it manually (Not sure but i ran `pip-compile requirements.dev.in`)

rdimaio commented on requirements/requirements.dev.in:
```diff
@@ -17,3 +15,4 @@ apispec-webframeworks                                       # Integration of api
 pip-tools==7.4.1                                            # Generate requirements files with pinned dependencies
 atlas-rucio-policy-package==0.4.0
 belleii-rucio-policy-package @ git+https://github.com/rucio/temporary-belle2-policy-package@v0.1.0
+ruff==0.9.10 # linter
```
nit: it would be nice to have this comment aligned with the other comments in the file, as well as having the `L` capitalized to match the other comments

rdimaio commented on tools/test/check_syntax.sh:
```diff
@@ -14,58 +14,27 @@
 echo '==============================='
 
-pylint --rcfile=pylintrc --errors-only $SYNTAX_PYLINT_ARGS
+ruff check --extend-ignore=E501,E722 $SYNTAX_RUFF_ARGS
```
Should we just include these ignored errors in the main Ruff config?

0xquark replied:
I think you're seeing an older commit here! I fixed this in the next [commit](https://github.com/rucio/rucio/pull/7481/commits/8906c23585348bcc58e764d1d88f50e4a923b69e).

rdimaio commented on tools/test/check_syntax.sh:
```diff
@@ -14,58 +14,27 @@
 
-    pylint --rcfile=pylintrc --reports y --exit-zero $SYNTAX_PYLINT_ARGS
-fi
+    ruff check --output-format=full $SYNTAX_RUFF_ARGS
```
Earlier we run ruff with `--extend-ignore=E501,E722`, so this syntax report would be inconsistent with the check done beforehand (extra reason to add them in the main Ruff config instead)

0xquark replied:
Answered in https://github.com/rucio/rucio/pull/7481#discussion_r1999446243

rdimaio commented on tools/test/ignoretool.py:
```diff
@@ -29,33 +29,37 @@
+    exclude_dirs = []
 
-    if script_args.flake8:
-        print(' '.join(map(lambda d: d + '/*.py', package_directories(PACKAGES))))
-    elif script_args.pylint:
-        ignore_dirs = []
-        include_dirs = set(package_directories(PACKAGES))
+    for root, dirs, files in os.walk('lib/'):
+        if '__pycache__' in root or root == 'lib/' or root in include_dirs:
+            continue
+        exclude_dirs.append(root)
 
-        for root, dirs, files in os.walk('lib/'):
-            if '__pycache__' not in root and root != 'lib/' and root not in include_dirs:
-                ignore_dirs.append(root)
```
I'm confused as why this was changed - the behaviour seems to be the same

0xquark replied:
I agree that the behavior is similar, but before, we were using `if not condition: do something` to check which directories to ignore. This made the logic a bit harder to follow. Now, we use `if condition: continue`, which immediately skips unwanted directories and only adds the valid ones. There’s no major performance difference, but this approach makes the code a little bit more cleaner and readable. Instead of checking what should be included, we now focus on filtering out what should be excluded first.

rdimaio commented on requirements/requirements.dev.txt:
```diff
@@ -6,21 +6,21 @@
     #   aiohttp
 alembic==1.14.1
-    # via -r requirements.server.txt
+    # via -r /root/PycharmProjects/rucio/requirements/requirements.server.txt
```
seems `pip-compile` has included your absolute path instead of relative path during compilation here - please revert

rdimaio reviewed (commented):
LGTM, just a small comment. Could you also please rebase to address the merge conflict?

rdimaio commented on pyproject.toml:
```diff
@@ -61,10 +61,13 @@ select = [
     "TCH", # flake8-type-checking
     "UP", # pyupgrade
     "TID", # flake8-tidy-imports
+    "PLE", #pylint
```
nit:
```suggestion
    "PLE", # pylint
```

rdimaio replied:
You can ignore this comment - we should do this in an automated way via a formatter at a later point
